File: frontend/lib/models/student.dart
import 'package:smart_insti_app/models/achievement.dart';
import 'package:smart_insti_app/models/skills.dart';

class Student {
  final String id;
  final String name;
  final String email;
  final String rollNumber;
  final String? about;
  final String? profilePicURI;
  final String? branch;
  final int? graduationYear;
  final List<Skill>? skills;
  final List<Achievement>? achievements;
  final List<String>? roles;
  final List<String> wishlist;

  Student({
    required this.id,
    required this.name,
    required this.email,
    required this.rollNumber,
    this.about,
    this.profilePicURI,
    this.branch,
    this.graduationYear,
    this.skills,
    this.achievements,
    this.roles,
    this.wishlist = const [],
  });

  factory Student.fromJson(Map<String, dynamic> json) {
    final academicInfo = json['academicInfo'] as Map<String, dynamic>?;
    return Student(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? 'Smart Insti User',
      email: json['email'] ?? '',
      rollNumber: json['rollNumber'] ?? json['student_id'] ?? academicInfo?['studentId'] ?? '',
      about: json['about'],
      profilePicURI: json['profilePicURI'] ?? json['profile_pic_url'],
      branch: json['branch'] ?? json['department'] ?? academicInfo?['department'],
      graduationYear: json['graduationYear'] ?? json['batch'] ?? academicInfo?['batch'],
      skills: json['skills'] != null
          ? (json['skills'] as List<dynamic>)
              .map((item) =>
                  item is Map<String, dynamic> ? Skill.fromJson(item) : null)
              .whereType<Skill>()
              .toList()
          : [],
      achievements: json['achievements'] != null
          ? (json['achievements'] as List<dynamic>)
              .map((item) => item is Map<String, dynamic>
                  ? Achievement.fromJson(item)
                  : null)
              .whereType<Achievement>()
              .toList()
          : null,
      roles: json['roles'] != null
          ? (json['roles'] as List?)?.map((item) => item as String).toList()
          : json['role'] != null ? [json['role'] as String] : [],
      wishlist: json['wishlist'] != null
          ? (json['wishlist'] as List?)
                  ?.map(
                      (item) => item is String ? item : item['_id'].toString())
                  .toList() ??
              []
          : [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'email': email,
      'rollNumber': rollNumber,
      'about': about,
      'profilePicURI': profilePicURI,
      'branch': branch,
      'graduationYear': graduationYear,
      'skills':
          skills != null ? skills!.map((skill) => skill.toJson()).toList() : [],
      'achievements': achievements != null
          ? achievements!.map((achievement) => achievement.toJson()).toList()
          : [],
      'roles': roles,
      'wishlist': wishlist,
    };
  }
}

File: backend/models/student.js
import mongoose from "mongoose";

const studentSchema = new mongoose.Schema(
  {
    name: { type: String, default: "Smart Insti User", trim: true },
    email: {
      type: String,
      required: true,
      unique: true,
      trim: true,
      index: true,
    },
    rollNumber: { type: String, unique: true, sparse: true, trim: true }, // Allows null but ensures uniqueness
    about: { type: String, trim: true },
    profilePicURI: { type: String, trim: true },
    branch: { type: String, trim: true },
    graduationYear: { type: Number, min: 1900, max: 2100 }, // Ensures valid graduation year
    skills: [{ type: mongoose.Schema.Types.ObjectId, ref: "Skill" }],
    achievements: [
      { type: mongoose.Schema.Types.ObjectId, ref: "Achievement" },
    ],
    wishlist: [{ type: mongoose.Schema.Types.ObjectId, ref: "Listing" }],
    fcmTokens: [{ type: String }],
    roles: { type: [String], default: [] }, // Ensures array structure
    currentOrganization: { type: String, trim: true }, // For Alumni
    designation: { type: String, trim: true }, // For Alumni
    linkedInProfile: { type: String, trim: true }, // For Alumni
    settings: {
        privacy: {
            showEmail: { type: Boolean, default: true },
            showPhone: { type: Boolean, default: false }, // If phone added later
            showAchievements: { type: Boolean, default: true }
        },
        notifications: {
            email: { type: Boolean, default: true },
            push: { type: Boolean, default: true }
        }
    }
  },
  { timestamps: true } // Adds createdAt & updatedAt fields
);

studentSchema.index({ name: "text", branch: "text", rollNumber: "text" });

const Student = mongoose.model("Student", studentSchema);
export default Student;
